src/shogun/mathematics/graph/nodes/BinaryNode.h

```diff
@@ -18,16 +18,36 @@ namespace shogun
 	{
 		namespace node
 		{
-			IGNORE_IN_CLASSLIST class BinaryNode : public Node
+			class BinaryNode;
+			class LogicalBinaryNode;
+
+			IGNORE_IN_CLASSLIST class BaseBinaryNode : public Node
 			{
+
+				friend class BinaryNode;
+				friend class LogicalBinaryNode;
+
 			public:
-				BinaryNode(
+				enum class BinaryShapeCompatibity
+				{
+					ArrayArray = 0,
+					ArrayScalar = 1,
+					BroadcastAlongAxis = 2
```

Comment:
One day this can be generalised to higher dimension tensors, but for now we just need scalar, vector and matrix working